forgerock-openbanking-uk-aspsp-as/forgerock-openbanking-uk-aspsp-as-gateway/forgerock-openbanking-uk-aspsp-as-gateway-server/src/main/java/com/forgerock/openbanking/aspsp/as/api/accesstoken/AccessTokenApiController.java

```diff
@@ -108,121 +109,32 @@ public ResponseEntity getAccessToken(MultiValueMap paramMap, String authorizatio
                 //proxy to AM
                 ResponseEntity responseFromAM = amGateway.toAM(request, new HttpHeaders(),
                         new ParameterizedTypeReference<AccessTokenResponse>() {}, paramMap);
-                responseEntity = ResponseEntity.status(responseFromAM.getStatusCode()).body(responseFromAM.getBody());
-        }
-
-        //Rewriting the response as we need to re-sign the id token
-        if (responseEntity.getStatusCode() == HttpStatus.OK) {
-            AccessTokenResponse accessTokenResponse = (AccessTokenResponse) responseEntity.getBody();
-            if (accessTokenResponse.getIdToken() != null) {
-                responseEntity = rewriteIdToken(accessTokenResponse);
-            }
-            tokenUsageService.incrementTokenUsage(accessTokenResponse);
+                log.debug("getAccessToken() response from AM; {}", responseFromAM);
+                if(!responseFromAM.getStatusCode().is2xxSuccessful() && !responseFromAM.getStatusCode().is3xxRedirection()){
+                    log.warn("getAccessToken() Un-successful call to AM to get access token. Status code: {}, body " +
+                            "{}",responseFromAM.getStatusCode(), responseFromAM.getBody() );
+                    throw new OBErrorResponseException(
+                            OBRIErrorType.ACCESS_TOKEN_INVALID.getHttpStatus(),
+                            OBRIErrorResponseCategory.ACCESS_TOKEN,
+                            OBRIErrorType.ACCESS_TOKEN_INVALID.toOBError1(responseFromAM.getBody()));
+                }
+                AccessTokenResponse accessTokenResponse = (AccessTokenResponse) responseFromAM.getBody();
+                log.debug("getAccessToken() accessTokenResponse from AM is {}", accessTokenResponse);
+                responseEntity = ResponseEntity.status(responseFromAM.getStatusCode()).body(accessTokenResponse);
```

Comment:
I don't understand why in this case a new ResponseEntity is created and returned, where as on line 101 above it is not. Could this be because in this latter case there is an async process involving UIs so the AM response data may no longer exist? Odd that the responses are treated so differently. 

I'd like to make the handling the same, but have not had time to investigate what an appropriate action might be here.. create a new ResponseEntity or return the ResponseEntity created from the call to AM.

Reply:
Looks like headLessAccessTokenService.getAccessToken creates a copy and returned a new ResponseEntity, similar to what is happening here. I could move the logic that tests for success or redirect into the headlessAccessTokenService, and also wrap some of this logic into a service too, which would tidy this method up and make it read better.

Reply:
It's up to you. Personally I wouldn't spend too much longer on it :)

Reply:
> I'm not sure either, but I wonder if it's because the response from AM contains headers we don't want to return?
> 
> As an aside, I'm finding the lack of type safety in the responses a bit confusing/frustrating. You've correctly defined `ResponseEntity<AccessTokenResponse>` above (though this method still returns just `ResponseEntity`), but `AMGateway` (which I know you didn't write) returns `ResponseEntity` without a generic type, so you need the cast to `AccessTokenResponse`. 🤷🏻‍♂️

I know. This is driving me spare. Where do I stop? Also when there is an error AM doesn't return an AccessTokenResponse, it returns some kind of error object. When I went to far with the parameterisation I got in trouble when there were error conditions. Really this needs better handling around the calls to AM before it can be fixed. I think that should probably be a different PR though.